### ISSUE TITLE
Dependency file not created when using -Wp,-MD,file.d,-MT,file.o

### ISSUE THREAD

**Issue by spstarr** [ccache/ccache#133]
When trying to build kernel a 2nd time I get this error:

If I use CCACHE_RECACHE=1 it works, but nothing is cached.

ccache version: 3.3.1

```
cat: /home/spstarr/rpmbuild/BUILD/kernel-4.8.0_rc7+/tools/objtool/.fixdep.o.d: No such file or directory
/home/spstarr/rpmbuild/BUILD/kernel-4.8.0_rc7+/tools/build/Makefile.build:77: recipe for target '/home/spstarr/rpmbuild/BUILD/kernel-4.8.0_rc7+/tools/objtool/fixdep.o' failed
make[6]: *** [/home/spstarr/rpmbuild/BUILD/kernel-4.8.0_rc7+/tools/objtool/fixdep.o] Error 1
Makefile:36: recipe for target '/home/spstarr/rpmbuild/BUILD/kernel-4.8.0_rc7+/tools/objtool/fixdep-in.o' failed
make[5]: *** [/home/spstarr/rpmbuild/BUILD/kernel-4.8.0_rc7+/tools/objtool/fixdep-in.o] Error 2
/home/spstarr/rpmbuild/BUILD/kernel-4.8.0_rc7+/tools/build/Makefile.include:7: recipe for target 'fixdep' failed
make[4]: *** [fixdep] Error 2
Makefile:60: recipe for target 'objtool' failed
make[3]: *** [objtool] Error 2
Makefile:1594: recipe for target 'tools/objtool' failed
make[2]: *** [tools/objtool] Error 2
make[2]: *** Waiting for unfinished jobs....
error: Bad exit status from /var/tmp/rpm-tmp.XJXuNg (%build)
```

**Comment by jpoimboe** [comment]
I see this issue too, just building the kernel from source (no RPM) with the O=<dir> kernel build option (which builds object files in a separate directory).

@torvalds also reported seeing it.

**Comment by jpoimboe** [comment]
I'm running Fedora 24.  When I downgrade from ccache-3.3.1-1.fc24.x86_64 to ccache-3.2.4-2.fc24.x86_64, it works fine again.

**Comment by jrosdahl** [comment]
> I see this issue too, just building the kernel from source (no RPM) with the O=dir kernel build option (which builds object files in a separate directory).

Do you have a more detailed reproduction recipe? I tried this:

```
make O=build defconfig
make O=build -j4
make O=build clean
make O=build -j4
```

Works fine for me with Linux bd5dbcb4be37ce97b115bf9e8d48ae75f0d8a9a2 and ccache 3.3.1 on Ubuntu 15.10 (with ccache masquerading as gcc).

**Comment by jpoimboe** [comment]
I think CONFIG_STACK_VALIDATION needs to be enabled.  Or just do:

```
make O=build allyesconfig
make O=build  # you can hit ctrl-C early on, after objtool gets built
make O=build clean
make O=build
```

**Comment by tanstafel** [comment]
Beside the occasional build errors there is another more hidden issue. Even if the kernel builds it will probably use old object code from the cache that does not represent the current source code. I just had a hard time finding out why my code changes did not do anything. Also the build time of the kernel in its output didn't change. Cleaning the cache helped. I also use ccache 3.3.1 on Fedora 24. This is really a critical bug!

**Comment by jpoimboe** [comment]
Slighter better repro procedure:
```
make O=build allyesconfig
make O=build tools/objtool
make O=build clean
make O=build tools/objtool
```

**Comment by jrosdahl** [comment]
> Slighter better repro procedure:

Excellent, thanks! Reproduced. It's a regression due to the fix of #129 (ccache 3.2.8 and 3.3.1 are affected).

**Comment by jrosdahl** [comment]
Fixed in ccache 3.2.9 and 3.3.2.
